cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java

```diff
@@ -100,7 +100,7 @@ public Watch createWatch(
       status.setLastWatchRequestTime(System.currentTimeMillis());
 
       Snapshot snapshot = snapshots.get(group);
-      String version = snapshot == null ? "" : snapshot.version(request.getTypeUrl());
+      String version = snapshot == null ? "" : snapshot.version(request.getTypeUrl(), request.getResourceNamesList());
```

Comment:
Hi

Replacing all occurences of snapshot.version(...)
with a new overridable method 
```
String version(Snapshot, DiscoveryRequest)
```
Would give the user a flexibility needed to implement custom versioning schemes

Reply:
Creating `version` method in `SimpleCache` public to let user override the version behaviour seems a bit hacky to me. What about a following compromise. Instead of `Snapshot#version(String typeUrl, List<String> resourceNames)` we can create `Snapshot#version(DiscoveryRequest)` so if user want to change this behaviour, he can provide his version of the snapshot with custom versioning without modifying the `SimpleCache`.

If the `SimpleCache` more appropriate place to put this logic then instead of making `version` method public I'd create a strategy `VersionProvider` or something like this, so user don't have to extends `SimpleCache`.